src/nvim/api/vim.c

```diff
@@ -790,14 +791,18 @@ void nvim_echo(Array chunks, Boolean history, Dict(echo_opts) *opts, Error *err)
     verbose_enter();
   }
 
-  msg_multihl(hl_msg, history ? "echomsg" : "echo", history);
+  if (opts->err) {
+    emsg_multiline(chunks.items[0].data.array.items[0].data.string.data, true);
+  } else {
+    msg_multihl(hl_msg, history ? "echomsg" : "echo", history);
+  }
 
   if (opts->verbose) {
     verbose_leave();
     verbose_stop();  // flush now
   }
 
-  if (history) {
+  if (history && !opts->err) {
     // history takes ownership
```

Comment:
is that the case even when "echo" is used on line 797?

Reply:
"echo" is used when `history == false`, so then hint in this comment doesn't apply either? Not sure what you mean, but yes when `history == false` `msg_multihl()` doesn't take ownership, and instead we free at the end of `nvim_echo()`.